sysfs/class_infiniband.go

```diff
@@ -0,0 +1,372 @@
+	path := fs.sys.Path(infinibandClassPath)
+
+	dirs, err := ioutil.ReadDir(path)
+	if err != nil {
+		return nil, fmt.Errorf("failed to list InfiniBand devices at %q: %v", path, err)
+	}
+
+	ibc := make(InfiniBandClass, len(dirs))
+	for _, d := range dirs {
+		device, err := fs.parseInfiniBandDevice(d.Name())
+		if err != nil {
+			return nil, err
+		}
+
+		ibc[device.Name] = *device
+	}
+
+	return ibc, nil
+}
+
+// Parse one InfiniBand device.
+func (fs FS) parseInfiniBandDevice(name string) (*InfiniBandDevice, error) {
+	path := fs.sys.Path(infinibandClassPath, name)
+	device := InfiniBandDevice{Name: name}
+
+	for _, f := range [3]string{"board_id", "fw_ver", "hca_type"} {
+		name := filepath.Join(path, f)
+		value, err := util.SysReadFile(name)
+		if err != nil {
+			return nil, fmt.Errorf("failed to read file %q: %v", name, err)
+		}
+
+		switch f {
+		case "board_id":
+			device.BoardID = value
+		case "fw_ver":
+			device.FirmwareVersion = value
+		case "hca_type":
+			device.HCAType = value
+		}
+	}
+
+	portsPath := filepath.Join(path, "ports")
+	ports, err := ioutil.ReadDir(portsPath)
+	if err != nil {
+		return nil, fmt.Errorf("failed to list InfiniBand ports at %q: %v", portsPath, err)
+	}
+
+	device.Ports = make(map[uint]InfiniBandPort, len(ports))
+	for _, d := range ports {
+		port, err := fs.parseInfiniBandPort(name, d.Name())
+		if err != nil {
+			return nil, err
+		}
+
+		device.Ports[port.Port] = *port
+	}
+
+	return &device, nil
+}
+
+// Parse InfiniBand state. Expected format: "<id>: <string-representation>"
+func parseState(s string) (uint, string, error) {
+	var id uint
+	var name string
+	parts := strings.Split(s, ":")
+	if len(parts) != 2 {
+		return id, name, fmt.Errorf("failed to split %s into 'ID: NAME'", s)
+	}
+	name = strings.TrimSpace(parts[1])
+	value, err := strconv.ParseUint(strings.TrimSpace(parts[0]), 10, 32)
+	if err != nil {
+		return id, name, fmt.Errorf("failed to convert %s into uint", strings.TrimSpace(parts[0]))
+	}
+	id = uint(value)
+	return id, name, nil
+}
+
+// Parse rate (example: "100 Gb/sec (4X EDR)") and return it as bytes/second
+func parseRate(s string) (uint64, error) {
+	var rate uint64
+	parts := strings.Split(s, "Gb/sec")
+	if len(parts) != 2 {
+		return rate, fmt.Errorf("failed to split '%s' by 'Gb/sec'", s)
+	}
+	value, err := strconv.ParseFloat(strings.TrimSpace(parts[0]), 32)
+	if err != nil {
+		return rate, fmt.Errorf("failed to convert %s into uint", strings.TrimSpace(parts[0]))
+	}
+	rate = uint64(value * 125000000)
+	return rate, nil
+}
+
+// parseInfiniBandPort scans predefined files in /sys/class/infiniband/<device>/ports/<port>
+// directory and gets their contents.
+func (fs FS) parseInfiniBandPort(name string, port string) (*InfiniBandPort, error) {
+	portNumber, err := strconv.ParseUint(port, 10, 32)
+	if err != nil {
+		return nil, fmt.Errorf("failed to convert %s into uint", port)
+	}
+	infiniBandPort := InfiniBandPort{Name: name, Port: uint(portNumber)}
```

Comment:
Nit: you can go with a shorter name since this is confined to a single function. Maybe `ibp`?

Reply:
Done.